File: CircularDoublyLinkedList.java
class Node {
    int data;
    Node prev, next;

    Node(int data) {
        this.data = data;
        this.prev = this.next = null;
    }
}

public class CircularDoublyLinkedList {
    Node head;

    // Insert node at the end
    void insertEnd(int data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
            newNode.next = newNode;
            newNode.prev = newNode;
            return;
        }

        Node tail = head.prev;
        tail.next = newNode;
        newNode.prev = tail;
        newNode.next = head;
        head.prev = newNode;
    }

    // Display the list
    void display() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }

        Node temp = head;
        System.out.print("Circular Doubly Linked List: ");
        do {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        } while (temp != head);
        System.out.println("(back to head)");
    }

    // Find the middle node
    void findMiddle() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }

        Node slow = head;
        Node fast = head;

        while (fast.next != head && fast.next.next != head) {
            slow = slow.next;
            fast = fast.next.next;
        }

        System.out.println("Middle value is: " + slow.data);
    }

    // Main method to test the code
    public static void main(String[] args) {
        CircularDoublyLinkedList cdll = new CircularDoublyLinkedList();

        for (int val : new int[] {10, 20, 30, 40, 50}) {
            cdll.insertEnd(val);
        }

        cdll.display();
        cdll.findMiddle();
    }
}

File: CircularDoublyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class CircularDoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            new_node.next = new_node
            new_node.prev = new_node
            return
        tail = self.head.prev
        tail.next = new_node
        new_node.prev = tail
        new_node.next = self.head
        self.head.prev = new_node

    def display(self):
        if not self.head:
            print("List is empty.")
            return
        temp = self.head
        print("Circular Doubly Linked List:", end=" ")
        while True:
            print(temp.data, end=" <-> ")
            temp = temp.next
            if temp == self.head:
                break
        print("(back to head)")

    def find_middle(self):
        if not self.head:
            print("List is empty.")
            return None

        slow = self.head
        fast = self.head

        while fast.next != self.head and fast.next.next != self.head:
            slow = slow.next
            fast = fast.next.next

        print("Middle value is:", slow.data)
        return slow.data

# Example usage
cdll = CircularDoublyLinkedList()
for val in [10, 20, 30, 40, 50]:
    cdll.insert_end(val)

cdll.display()
cdll.find_middle()
